internal/compattest/compattest_test.go

```diff
@@ -0,0 +1,238 @@
+}
+
+func (s *compatTestSuiteBase) replayPCRSequenceFromReader(c *C, r io.Reader) {
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		components := strings.Split(scanner.Text(), " ")
+		c.Assert(len(components), Equals, 3)
+
+		pcr, err := strconv.Atoi(components[0])
+		c.Assert(err, IsNil)
+		alg, err := strconv.ParseUint(components[1], 10, 16)
+		c.Assert(err, IsNil)
+		digest, err := hex.DecodeString(components[2])
+		c.Assert(err, IsNil)
+
+		c.Assert(s.TPM.PCRExtend(s.TPM.PCRHandleContext(pcr), tpm2.TaggedHashList{{HashAlg: tpm2.HashAlgorithmId(alg), Digest: digest}}, nil), IsNil)
+	}
+}
+
+func (s *compatTestSuiteBase) replayPCRSequenceFromFile(c *C, path string) {
+	f, err := os.Open(path)
+	c.Assert(err, IsNil)
+	defer f.Close()
+
+	s.replayPCRSequenceFromReader(c, f)
+}
+
+func (s *compatTestSuiteBase) testUnseal(c *C, pcrEventsFile string) {
+	s.replayPCRSequenceFromFile(c, pcrEventsFile)
+
+	k, err := secboot.ReadSealedKeyObject(s.absPath("key"))
+	c.Assert(err, IsNil)
+	c.Check(k.AuthMode2F(), Equals, secboot.AuthModeNone)
+
+	key, err := k.UnsealFromTPM(s.TPM, "")
+	c.Check(err, IsNil)
+
+	expectedKey, err := ioutil.ReadFile(s.absPath("clearKey"))
+	c.Assert(err, IsNil)
+
+	c.Check(key, DeepEquals, expectedKey)
+}
+
+func (s *compatTestSuiteBase) TestChangePIN(c *C) {
+	k, err := secboot.ReadSealedKeyObject(s.absPath("key"))
+	c.Assert(err, IsNil)
+	c.Check(k.AuthMode2F(), Equals, secboot.AuthModeNone)
+
+	c.Check(secboot.ChangePIN(s.TPM, s.absPath("key"), "", testPIN), IsNil)
+	k, err = secboot.ReadSealedKeyObject(s.absPath("key"))
+	c.Assert(err, IsNil)
+	c.Check(k.AuthMode2F(), Equals, secboot.AuthModePIN)
+
+	c.Check(secboot.ChangePIN(s.TPM, s.absPath("key"), testPIN, ""), IsNil)
+	k, err = secboot.ReadSealedKeyObject(s.absPath("key"))
+	c.Assert(err, IsNil)
+	c.Check(k.AuthMode2F(), Equals, secboot.AuthModeNone)
+}
+
+func (s *compatTestSuiteBase) testUnsealWithPIN(c *C, pcrEventsFile string) {
+	c.Check(secboot.ChangePIN(s.TPM, s.absPath("key"), "", testPIN), IsNil)
+
+	s.replayPCRSequenceFromFile(c, pcrEventsFile)
+
+	k, err := secboot.ReadSealedKeyObject(s.absPath("key"))
+	c.Assert(err, IsNil)
+	c.Check(k.AuthMode2F(), Equals, secboot.AuthModePIN)
+
+	key, err := k.UnsealFromTPM(s.TPM, testPIN)
+	c.Check(err, IsNil)
+
+	expectedKey, err := ioutil.ReadFile(s.absPath("clearKey"))
+	c.Assert(err, IsNil)
+
+	c.Check(key, DeepEquals, expectedKey)
+}
+
+func (s *compatTestSuiteBase) testUpdateKeyPCRProtectionPolicy(c *C, pcrProfile *secboot.PCRProtectionProfile) {
+	c.Check(secboot.UpdateKeyPCRProtectionPolicy(s.TPM, s.absPath("key"), s.absPath("pud"), pcrProfile), IsNil)
+}
+
+func (s *compatTestSuiteBase) testUpdateKeyPCRProtectionPolicyRevokes(c *C, pcrProfile *secboot.PCRProtectionProfile, pcrEventsFile string) {
+	tmpDir := c.MkDir()
+	keyFile, err := os.Open(s.absPath("key"))
+	c.Assert(err, IsNil)
+	defer keyFile.Close()
+
+	keyFileBackup, err := os.Create(filepath.Join(tmpDir, "key"))
```

Comment:
What is the purpose of creating this keyfile backup? It seems that it's not used in this test.

Reply:
It's loaded a few lines down:

https://github.com/snapcore/secboot/pull/86/files/962267b8cad3a3d0c4019290b60572bee4c1c590#diff-1c58d1fcc056a4f05e5d823f6b670648R204

Reply:
Ah right. Thanks.